lib/experimental/html/class-wp-html-tag-processor.php

```diff
@@ -1378,50 +1380,66 @@ private static function sort_start_ascending( $a, $b ) {
 	}
 
 	/**
-	 * Given an attribute name, return its value as set in $lexical_updates.
+	 * Return the enqueued value for a given attribute, if one exists.
+	 *
+	 * Enqueued updates can take different data types:
+	 *  - If an update is enqueued and is boolean, the return will be `true`
+	 *  - If an update is otherwise enqueued, the return will be the string value of that update.
+	 *  - If an attribute is enqueued to be removed, the return will be `null` to indicate that.
+	 *  - If no updates are enqueued, the return will be `false` to differentiate from "removed."
```

Comment:
The semantics of this are a bit tricky 🤔 It's probably inherent to the fact that on top of HTML attribute semantics (which are reflected by`get_attribute`), we also need a value to mean "no updates", but I'm not sure if `false` would be my preferred choice. (I was a bit confused when first reading conditionals like [this](https://github.com/WordPress/gutenberg/pull/46680/commits/fafafce0d5d667a4a883ccb12bebd3fb03b1e59d#diff-93724b2f6e5dbf16c2b3c0efea57fa442b572f75f325d1cb185cf614b692cf9fR1135).)

(Admittedly, I can't think of a much better semantics off the top of my head.)

Reply:
I left the commit with the thought of adding class constants which would reflect better. Left them as a follow-up to be sure it's a good idea.

If `null` weren't already used to mean "no attribute" I'd use that here, but in the case of updates, we can _update_ and attribute to _remove it_ which would indicate `null` 😆 

Thankfully `false` is not possible as an attribute value so I reached for it out of convenience.

The one issue with constants is that we have to be careful not to conflate them with legitimate attribute values, so strings are out, leaving us back with `null` and `false`. We could also create empty classes as a faux enum, but then we're adding potentially needless complexity and overhead.

Reply:
Yeah, agree that `false` seems like a natural contender, since it was available. It's still not the most intuitive thing, but since this is a private method, and you've nicely documented it, let's go with it 👍

Reply:
@dmsnell constants like ‚ATTR_REMOVE = null’ or ‚ATTR_NO_UPDATES = null’ would make this easier to reason about AND preserve the semantics you described

Reply:
that's one thing I thought of @adamziel but then the worry I had was how those obscure the values. unlike with `ADD_CLASS` and `REMOVE_CLASS`, we actually use the return values here and get no type safety.

for example, this code looks clear…

```php
if ( self::NO_UPDATES === $this->get_enqueued_attribute_value( $name ) ) {
	…
}
```

but then because these constants look like real expected values I think there's some obscurity in this code…

```php
if ( null === $this->get_enqueuedd_attribute_value( $name ) ) {
	…
}
```

just noting here that I had some reticence in jumping to constants because of the way they hide the values that we do end up using as those primitive values.